File: ios/Runner/SpeechLocalizerPlugin.swift
import Foundation
import Flutter
import Accelerate

@objc class SpeechLocalizerPlugin: NSObject, FlutterPlugin {
    static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "live_captions_xr/speech_localizer", binaryMessenger: registrar.messenger())
        let instance = SpeechLocalizerPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "estimateDirection":
            guard let args = call.arguments as? [String: Any],
                  let left = args["left"] as? FlutterStandardTypedData,
                  let right = args["right"] as? FlutterStandardTypedData,
                  let sampleRate = args["sampleRate"] as? Double else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Missing or invalid arguments", details: nil))
                return
            }
            let micDistance = args["micDistance"] as? Double ?? 0.08
            let soundSpeed = args["soundSpeed"] as? Double ?? 343.0
            let angle = Self.estimateDirectionGccPhat(
                left: left.data,
                right: right.data,
                sampleRate: sampleRate,
                micDistance: micDistance,
                soundSpeed: soundSpeed
            )
            result(angle)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    static func estimateDirectionGccPhat(left: Data, right: Data, sampleRate: Double, micDistance: Double, soundSpeed: Double) -> Double {
        let n = min(left.count, right.count) / 4
        guard n > 0 else { return 0 }
        let N = 1 << (Int(log2(Double(n))) + 1)
        var leftFloats = [Float](repeating: 0, count: N)
        var rightFloats = [Float](repeating: 0, count: N)
        left.withUnsafeBytes { ptr in
            _ = memcpy(&leftFloats, ptr.baseAddress!, n * 4)
        }
        right.withUnsafeBytes { ptr in
            _ = memcpy(&rightFloats, ptr.baseAddress!, n * 4)
        }
        // Zero-pad
        for i in n..<N { leftFloats[i] = 0; rightFloats[i] = 0 }
        var splitLeft = DSPSplitComplex(realp: .allocate(capacity: N/2), imagp: .allocate(capacity: N/2))
        var splitRight = DSPSplitComplex(realp: .allocate(capacity: N/2), imagp: .allocate(capacity: N/2))
        let fftSetup = vDSP_create_fftsetup(vDSP_Length(log2(Double(N))), FFTRadix(kFFTRadix2))
        leftFloats.withUnsafeBufferPointer { lPtr in
            rightFloats.withUnsafeBufferPointer { rPtr in
                vDSP_ctoz(UnsafePointer<DSPComplex>(OpaquePointer(lPtr.baseAddress!)), 2, &splitLeft, 1, vDSP_Length(N/2))
                vDSP_ctoz(UnsafePointer<DSPComplex>(OpaquePointer(rPtr.baseAddress!)), 2, &splitRight, 1, vDSP_Length(N/2))
                vDSP_fft_zrip(fftSetup!, &splitLeft, 1, vDSP_Length(log2(Double(N))), FFTDirection(FFT_FORWARD))
                vDSP_fft_zrip(fftSetup!, &splitRight, 1, vDSP_Length(log2(Double(N))), FFTDirection(FFT_FORWARD))
                // GCC-PHAT cross spectrum
                var crossReal = [Float](repeating: 0, count: N/2)
                var crossImag = [Float](repeating: 0, count: N/2)
                for i in 0..<N/2 {
                    let lr = splitLeft.realp[i] * splitRight.realp[i] + splitLeft.imagp[i] * splitRight.imagp[i]
                    let li = splitLeft.imagp[i] * splitRight.realp[i] - splitLeft.realp[i] * splitRight.imagp[i]
                    let mag = sqrt(lr * lr + li * li)
                    if mag > 1e-8 {
                        crossReal[i] = lr / mag
                        crossImag[i] = li / mag
                    }
                }
                var cross = DSPSplitComplex(realp: &crossReal, imagp: &crossImag)
                vDSP_fft_zrip(fftSetup!, &cross, 1, vDSP_Length(log2(Double(N))), FFTDirection(FFT_INVERSE))
                // Find max index (time delay)
                var corr = [Float](repeating: 0, count: N)
                corr.withUnsafeMutableBufferPointer { corrPtr in
                    corrPtr.baseAddress!.withMemoryRebound(to: DSPComplex.self, capacity: N/2) { complexPtr in
                        vDSP_ztoc(&cross, 1, complexPtr, 2, vDSP_Length(N/2))
                    }
                }
                var maxVal: Float = 0
                var maxIdx: vDSP_Length = 0
                vDSP_maxvi(corr, 1, &maxVal, &maxIdx, vDSP_Length(N))
                var delay = Int(maxIdx)
                if delay > N/2 { delay = delay - N }
                let timeDelay = Double(delay) / sampleRate
                let maxDelay = micDistance / soundSpeed
                let clamped = max(-1.0, min(timeDelay / maxDelay, 1.0))
                vDSP_destroy_fftsetup(fftSetup)
                splitLeft.realp.deallocate(); splitLeft.imagp.deallocate()
                splitRight.realp.deallocate(); splitRight.imagp.deallocate()
                return Double(asin(clamped))
            }
        }
        return 0
    }
} 

File: android/app/src/main/kotlin/com/example/aurasense/SpeechLocalizerPlugin.kt
package com.example.aurasense

import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.PluginRegistry.Registrar
import org.jtransforms.fft.FloatFFT_1D
import kotlin.math.*

class SpeechLocalizerPlugin : MethodCallHandler {
    companion object {
        @JvmStatic
        fun registerWith(registrar: Registrar) {
            val channel = MethodChannel(registrar.messenger(), "live_captions_xr/speech_localizer")
            channel.setMethodCallHandler(SpeechLocalizerPlugin())
        }
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "estimateDirection" -> {
                val args = call.arguments as? Map<*, *> ?: run {
                    result.error("INVALID_ARGUMENT", "Missing arguments", null)
                    return
                }
                val left = (args["left"] as? ByteArray)?.toFloatArray()
                val right = (args["right"] as? ByteArray)?.toFloatArray()
                val sampleRate = (args["sampleRate"] as? Double) ?: 16000.0
                val micDistance = (args["micDistance"] as? Double) ?: 0.08
                val soundSpeed = (args["soundSpeed"] as? Double) ?: 343.0
                if (left == null || right == null) {
                    result.error("INVALID_ARGUMENT", "Missing left or right channel", null)
                    return
                }
                val angle = estimateDirectionGccPhat(left, right, sampleRate, micDistance, soundSpeed)
                result.success(angle)
            }
            else -> result.notImplemented()
        }
    }

    private fun ByteArray.toFloatArray(): FloatArray {
        val floats = FloatArray(this.size / 4)
        for (i in floats.indices) {
            val bits = (this[i * 4].toInt() and 0xFF) or
                    ((this[i * 4 + 1].toInt() and 0xFF) shl 8) or
                    ((this[i * 4 + 2].toInt() and 0xFF) shl 16) or
                    ((this[i * 4 + 3].toInt() and 0xFF) shl 24)
            floats[i] = Float.fromBits(bits)
        }
        return floats
    }

    private fun estimateDirectionGccPhat(
        left: FloatArray, right: FloatArray,
        sampleRate: Double, micDistance: Double, soundSpeed: Double
    ): Double {
        val n = min(left.size, right.size)
        val N = 1 shl (ceil(log2(n.toDouble())).toInt())
        val leftPadded = left.copyOf(N)
        val rightPadded = right.copyOf(N)
        val leftFFT = FloatArray(2 * N)
        val rightFFT = FloatArray(2 * N)
        for (i in 0 until N) {
            leftFFT[2 * i] = leftPadded[i]
            rightFFT[2 * i] = rightPadded[i]
        }
        val fft = FloatFFT_1D(N.toLong())
        fft.realForwardFull(leftFFT)
        fft.realForwardFull(rightFFT)
        val cross = FloatArray(2 * N)
        for (i in 0 until N) {
            val lr = leftFFT[2 * i] * rightFFT[2 * i] + leftFFT[2 * i + 1] * rightFFT[2 * i + 1]
            val li = leftFFT[2 * i + 1] * rightFFT[2 * i] - leftFFT[2 * i] * rightFFT[2 * i + 1]
            val mag = sqrt(lr * lr + li * li)
            if (mag > 1e-8) {
                cross[2 * i] = lr / mag
                cross[2 * i + 1] = li / mag
            }
        }
        fft.complexInverse(cross, true)
        var maxVal = cross[0]
        var maxIdx = 0
        for (i in 1 until N) {
            if (cross[2 * i] > maxVal) {
                maxVal = cross[2 * i]
                maxIdx = i
            }
        }
        var delay = maxIdx
        if (delay > N / 2) delay -= N
        val timeDelay = delay / sampleRate
        val maxDelay = micDistance / soundSpeed
        val clamped = timeDelay / maxDelay
        return asin(clamped.coerceIn(-1.0, 1.0))
    }
} 